### ISSUE TITLE
[@walletconnect/ethereum-provider V2] Intermittent `Error: No matching key.` on `eth_sendTransaction` requests

### ISSUE THREAD

**Issue by JesseKuntz** [WalletConnect/walletconnect-monorepo#2326]
**Describe the bug**
When sending an `eth_sendTransaction` request, seeing intermittent `Error: No matching key.` errors on both the `history` and `keychain` keys, both locally and in a hosted environment. Sometimes they work, sometimes not. For example, I've gotten it to work twice in a row and then on the 3rd time we'll get the error. The result is that we never get a response from the request - it just hangs. Additional info at the bottom.

**SDK Version (if relevant)**
 - Client: `@walletconnect/ethereum-provider`
 - Version: `2.7.2`

**To Reproduce**
1. Initialize a provider using `@walletconnect/ethereum-provider` `init` 
2. Connect an account using the provider
3. Fire a send tx request:
```
provider.request<string>({
    method: "eth_sendTransaction",
    params: [{...}],
});
```
5. NEVER get a response back from the request.

**Expected behavior**
The `eth_sendTransaction` requests succeed every time.

**Screenshots**

![image](https://user-images.githubusercontent.com/8867651/236959211-84eb7b5f-1a92-46c2-996f-7c0886717af5.png)

**Desktop (please complete the following information):**
 - OS: macOS
 - Browser: Chrome
 - Version: 112.0.5615.121

**Additional context**

Here's the WalletConnect local storage `history` entry for the error you see in the screenshot:

```
{
    "id": 1683064859400268,
    "topic": "36ec8902c543226afaa2c6c9978b7517576718a6339bf389275d18a779e4b335",
    "request": {
        "method": "wc_sessionRequest",
        "params": {
            "request": {
                "method": "eth_sendTransaction",
                "params": [
                    {
                        "from": REDACTED,
                        "to": REDACTED,
                        "value": "0x1bc16d674ec800000",
                        "data":  REDACTED
                    }
                ]
            },
            "chainId": "eip155:5"
        }
    },
    "chainId": null
}
```

As you can see, there's no `response` key in that entry like there is for requests that get a response.

**Comment by anuj16102062** [comment]
any update on this

**Comment by finessevanes** [comment]
@anuj16102062 pinging our team about this.

**Comment by Pasha8914** [comment]
have same issue, after network switch

https://github.com/orgs/WalletConnect/discussions/2456#discussioncomment-6008951

**Comment by ganchoradkov** [comment]
Hey @JesseKuntz the `no matching key: history` error indicates that the SDK receives a response for a request it didn't sent, or it did but the ID is missing in the history. 
Do you have any custom logic modifying or clearing the localStorage?

**Comment by ganchoradkov** [comment]
I managed to reproduce it only if I 
- manually nuke the history `wc@2:core:0.3//history`
- reload the client
- approve the request in the wallet

**Comment by JesseKuntz** [comment]
Hey @ganchoradkov thanks for taking a look at this. I do have some logic to clear something out of the local storage, but it's the `WALLETCONNECT_DEEPLINK_CHOICE`, not the `wc@2:core:0.3//history` (there's actually another issue for that here: https://github.com/WalletConnect/walletconnect-monorepo/issues/2303#issuecomment-1539185204). I don't touch the `wc@2:core:0.3//history` at all.

I've gone ahead and upgraded `@walletconnect/ethereum-provider` to latest (2.8.0), but now I'm getting some other errors:

- When connecting (after approving the connection in the wallet), I'll get: 
`Error: No matching key. pairing: undefined`
But the connection looks to still work, there's just no entry in `wc@2:core:0.3//pairing`. Is that okay?
- When sending a transaction, but then receiving a thrown error from the wallet (expected, such as when canceling a transaction), I'll get:
`Error: No matching key. keychain: 5c432db6f7e4c08b3137b1fa9038988a8217dad60564266cc18ad59c6221a57c`
`Error: No matching key. history: 1686332171383594`

Here's a full example of the `history` one:
```
Uncaught (in promise) Error: No matching key. history: 1686333621898358
    at Ut.getRecord (webpack-internal:///(app-client)/../../node_modules/.pnpm/@walletconnect+core@2.8.0_2adebc2xdjqbcvbjxkodkhadp4/node_modules/@walletconnect/core/dist/index.es.js:82:58216)
    at Ut.get (webpack-internal:///(app-client)/../../node_modules/.pnpm/@walletconnect+core@2.8.0_2adebc2xdjqbcvbjxkodkhadp4/node_modules/@walletconnect/core/dist/index.es.js:82:56588)
    at Nt.onRelayEventResponse (webpack-internal:///(app-client)/../../node_modules/.pnpm/@walletconnect+core@2.8.0_2adebc2xdjqbcvbjxkodkhadp4/node_modules/@walletconnect/core/dist/index.es.js:82:50376)
    at EventEmitter.eval (webpack-internal:///(app-client)/../../node_modules/.pnpm/@walletconnect+core@2.8.0_2adebc2xdjqbcvbjxkodkhadp4/node_modules/@walletconnect/core/dist/index.es.js:82:54801)
```

Interestingly enough, both the `keychain` and `history` referenced appear in local storage. Could it be that there's some race condition in the SDK that checks for the `key` before the entry exists?

**Comment by mihaic195** [comment]
I'm having the same issue.
![Screenshot 2023-06-19 at 12 04 31](https://github.com/WalletConnect/walletconnect-monorepo/assets/34626017/86f9de4d-9bb9-4479-9cc1-a03c8ea46a03)

**Comment by siosio34** [comment]
I'm having the same issue. too

**Comment by JesseKuntz** [comment]
Some updates:

> When connecting (after approving the connection in the wallet), I'll get:
Error: No matching key. pairing: undefined
But the connection looks to still work, there's just no entry in wc@2:core:0.3//pairing. Is that okay?

☝🏻 Looks like this has been fixed in `2.8.2`

> When sending a transaction, but then receiving a thrown error from the wallet (expected, such as when canceling a transaction), I'll get:
Error: No matching key. keychain: 5c432db6f7e4c08b3137b1fa9038988a8217dad60564266cc18ad59c6221a57c
Error: No matching key. history: 1686332171383594

☝🏻 I've adjusted our usage of `@walletconnect/ethereum-provider` to be closer to how `wagmi` uses it, and I'm not seeing errors like this anymore. I think it was due to some re-initializations of the provider, when you _really_ only want to initialize it (and `.enable()` it) once (at least per page load). This was not super clear to me as in v1 of `@walletconnect/ethereum-provider` we had to re-initialize it in a few different cases.

@ganchoradkov if you think that this could be why those errors would appear intermittently, I think this can be closed.

**Comment by CSSE-WZ** [comment]
I'm having the same issue. too @JesseKuntz 
When I use the wallet to log in, as long as I refuse to log in and scan the code to log in again, this error will be reported:
![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/22407879/ce78e6a6-69d7-4ec4-bf88-b829b372dacf)

**Comment by equalogix** [comment]
I'm using `v2.8.6` and I am facing the same issue. In my case it occurs only after a page refresh. After connecting with the modal everything works, but if the page gets refreshed and I then do any any action (ie: sign message, disconnect), the action is completed, but I get the error `no matching key. History ....`. If I then clear the error, and then sign another message, it continues to send message correctly, but I get the error again.

I tried to catch the error, for example with `signer.signMessage(...).then(...).catch(err =>  ...)` but the error is not being caught as it appears to originate as a side effect of the `signMessage` call.

I also tried "nuking" the `history` in storage, but that didn't help.

Update: so far I had been running the `create-react-app` using the dev build `npm start`, I tried creating a production build to see how the error would be handled, and to my surprise there were no errors. It runs as expected.

**Comment by ylmin** [comment]
<img width="659" alt="image" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/87960263/d9d77198-d2e1-4983-8255-ff926c3d879e">

**Comment by xircusfounder** [comment]
Got the same error

![Screen Shot 2023-07-12 at 11 15 29 AM](https://github.com/WalletConnect/walletconnect-monorepo/assets/86594363/037c2227-f05d-43b8-b496-f8aaf1b8238c)

**Comment by frankye23** [comment]
I upgrade @walletconnect/ethereum-provider to 2.8.2, the error seems to continue exist. Luckily, it doesn't have influence on transaction, signing message, etc. It will only appear when i refuse the request from the dapp.

**Comment by 0xSamrat** [comment]
I am also getting the same problem, I stop using "...wallets" in "connectorsForWallets" and started using wallet connect itself and also putting the whole function call inside to try and catch block.

**Comment by Pavan-0xblock** [comment]
Same issue here also when I reject the request at that time I get the same.

![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/100664726/f8e56592-fc4d-4c69-bc17-dee9e406b7f2)

**Comment by glitch-txs** [comment]
Aswered [here](https://github.com/orgs/WalletConnect/discussions/3291). Seems to be an issue on the Wallet side.

**Comment by evan-metaseed** [comment]
I had issues related to this with local storage variables stored in my browser. Clearing them programmatically fixed this.

**Comment by moondayli** [comment]
> I had issues related to this with local storage variables stored in my browser. Clearing them programmatically fixed this.

how please share code

**Comment by evan-metaseed** [comment]
You can clear local storage in a browser with JS by doing `localStorage.clear()`

**Comment by glitch-txs** [comment]
This is an issue on the wallet side, which wallet are you using?

**Comment by evan-metaseed** [comment]
I was using wallet connect in conjunction with ronin wallet

**Comment by moondayli** [comment]
> This is an issue on the wallet side, which wallet are you using?

I'm using phantom wallet but some how I don't know how they did it but phantom wallet always popups no matter what I do the set provider function I got from github works for other wallets except phantom somehow they're able to beat metamask I have cleared my storage same thing, I can decide not to add phantom to my website but even at that if the user has a phantom wallet it will still connect to it and ignore metamask and Coinbase if anyone has a solution to this phantom issue pleaseeeeee let me know so I can fix it I have been trying to fix it for about three days now and no success, Opensea fixed it tho I have been trying to get their dev team to explain to me how but to success on that as well 🤧

**Comment by somedeveloper00** [comment]
In Unity, disable the "Resume Session On Init" if you're using the official prefab. You'll later have to manually handle session resuming I guess.

![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/79690923/cf77a6d5-67a9-41a7-b87d-28daa2d65436)
